Automattic/wp-calypso
# Checkout: Convert CompositeCheckout component to TypeScript

#### Changes proposed in this Pull Request

This converts the main component in new checkout, `CompositeCheckout`, to TypeScript. In the process, it fixes several small bugs and type errors.

#### Testing instructions

This touches nearly all aspects of checkout, but it's impossible to test them all, and like many other recent refactors, if anything serious is broken by this, it should be obvious quickly. 

Notable changes:

- `arguments` has been removed from the `THANK_YOU_URL_GENERATED` analytics event. They have actually been missing since https://github.com/Automattic/wp-calypso/pull/46423. It would be nice to add them back in, but getting them out of the hook is actually non-trivial and I'd rather that be in its own PR. For the moment, this just removes the unused property. This should have no effect, but it might be worth examining the event when it runs to be sure it's working.
- This moves`doesValueExist` to a top-level helper in the composite checkout directory which can be used as `[].filter(doesValueExist)` in place of `[].filter(Boolean)` to please TypeScript. This should have no user-facing effect as long as it filters out falsy values correctly. If it doesn't, we'd probably see lots of errors about adding products when no product add errors exist.
- This makes `useRedirectIfCartEmpty` more explicit about when it will and won't redirect; there's now a single argument, `doNotRedirect`, which is used for that decision, putting the logic fully in the consumer. Worth testing that we are still redirected when we'd expect to be, and are not redirected if the cart is empty and an invalid product slug is in the URL.
